Support for multipart/form-data in the request body

ilanashapiro opened the pull request:
Addresses #2008

**Summary**

Currently, validation always fails whenever a multipart/form-data content type is used in the request body. This prevents the normal usage of prism without any other proxy server (like nginx) on which such requests can be filtered out. Now, users can successfully use multipart/form-data content type in the request body.

**Checklist**

- The basics
  - [X] I tested these changes manually in my local or dev environment
- Tests
  - [X] Added or updated
  - [ ] N/A
- Event Tracking
  - [ ] I added event tracking and followed the event tracking guidelines
  - [X] N/A
- Error Reporting
  - [ ] I reported errors and followed the error reporting guidelines
  - [X] N/A

**Screenshots**

Given the following file called examples/testapi.yaml:
```
openapi: '3.1.0'
paths:
  /path:
    post:
      responses:
        200:
          content:
            text/plain:
              example: ok
      requestBody:
        content:
          multipart/form-data:
            schema:
              type: object
              properties:
                testData:
                  type: string
              required:
                - "testData"
```
Writing a non-empty post request to http://127.0.0.1/path with included form-data in the request body now produces the following, successfully returning the response defined in the example:
<img width="1003" alt="image" src="https://github.com/stoplightio/prism/assets/28958079/fdba4800-0446-49c9-b74f-0f7243f256bb">
<img width="808" alt="image" src="https://github.com/stoplightio/prism/assets/28958079/8f650b74-3b86-4d18-8997-da6b3862b2c1">

In contrast, running this same command on the old version of Prism gives a 422 error:
<img width="1003" alt="image" src="https://github.com/stoplightio/prism/assets/28958079/70088729-f0cb-48ed-aeae-9b77f9bb5528">
<img width="1012" alt="image" src="https://github.com/stoplightio/prism/assets/28958079/2afaef08-5dfe-4bdc-8f30-2bcc5e545288">

ilanashapiro commented:
@rattrayalex @RobertCraigie

chohmann commented on packages/http-server/src/__tests__/body-params-validation.spec.ts:
```diff
@@ -796,5 +860,20 @@ describe('body params validation', () => {
+          headers: { 'content-type': 'multipart/form-data' },
+        });
+
+        expect(response.status).toBe(200);
```
the test says it returns 422, but the assertion says it returns 200. What was the intent for this test? It seems like you're trying to test when passing an invalid multipart form data since the body is using URLSearchParams which is not multipart form data.

chohmann commented on test-harness/specs/validate-body-params/multipart-allow-empty-value-2.oas2.txt:
```diff
@@ -0,0 +1,28 @@
+====test====
+If I have an operation with a parameter with the allowEmptyValue set to false to false
```
```suggestion
If I have an operation with a parameter with the allowEmptyValue set to false
```

chohmann commented on test-harness/specs/validate-body-params/multipart-allow-empty-value-2.oas2.txt:
```diff
@@ -0,0 +1,28 @@
+====test====
+If I have an operation with a parameter with the allowEmptyValue set to false to false
+and I send a request with a correcly encoded parameter but with an empty key
```
```suggestion
and I send a request with a correctly encoded parameter but with an empty key
```

chohmann commented on test-harness/specs/validate-body-params/multipart-allow-empty-value.oas2.txt:
```diff
@@ -0,0 +1,30 @@
+====test====
+If I have an operation with a parameter with the allowEmptyValue set to false to false
```
```suggestion
If I have an operation with a parameter with the allowEmptyValue set to true
```

chohmann commented on test-harness/specs/validate-body-params/multipart-allow-empty-value.oas2.txt:
```diff
@@ -0,0 +1,30 @@
+====test====
+If I have an operation with a parameter with the allowEmptyValue set to false to false
+and I send a request with a correcly encoded parameter and with the parameter empty
```
```suggestion
and I send a request with a correctly encoded parameter and with the parameter empty
```

chohmann commented on test-harness/specs/validate-body-params/multipart-form-data-allow-reserved-ok.oas3.txt:
```diff
@@ -0,0 +1,34 @@
+====server====
+mock -p 4010 ${document}
+====command====
+curl -i -X POST http://localhost:4010/path -H "Content-Type: multipart/form-data" --data "reserved=:/?#[]@!$&'()*+,;"
```
For all of these curl requests you need to use `-F` instead of `--data` as that will actually pass form data as multipart.

chohmann reviewed (changes requested):
Thank you for submitting this PR!

The main issue we see with this PR is the test harness tests don't actually send multipart form data as expected.

Each of those tests needs to use the `-F` flag instead of `--data` in the curl command.

ilanashapiro commented:
@chohmann All requested changes are now implemented. Thank you for pointing out that I needed to pass data with the -F flag rather than the --data flag! I initially misunderstood that setting "Content-type: multipart/form-data" was not sufficient. And apologies about the typos in tests I added to the test harness -- I copied these tests from existing files in the application/x-www-form-urlencoded test suite and didn't proofread them first. I have now fixed all the typos in these as well as in the original files. I also added some regex parsing logic for multipart form data to handle additional cases. Please let me know if you'd like any further changes! Thanks!

EdVinyard commented on packages/http/src/validator/validators/body.ts:
```diff
@@ -54,6 +54,15 @@ export function splitUriParams(target: string) {
 }
 
+export function parseMultipartFormDataParams(target: string) {
+  const pairs = target.split(/--.*\s*.*="/).filter(element => element);
```
I think this code assumes that the boundary between multipart entities is `--` followed by _any_ string.  However, the boundary between multipart entities is defined, in each individual request, by the request's `Content-Type` header's `Boundary` parameter.  This is because the client must choose a value that doesn't occur _within_ any multipart entity that it will transmit.  For example, consider the effect of the following example on your implementation.

```
curl -v -F a=b -F c='--="' <hostname>
```

Also consider any form value that includes `\r\n\r\n`.  

[RFC 2046 § 5.1.1](https://datatracker.ietf.org/doc/html/rfc2046#section-5.1.1) describes the boundary as...

> a line consisting entirely of two hyphen characters ("-", decimal value 45) followed by the boundary parameter value from the Content-Type header field, optional linear whitespace, and a terminating CRLF.

I think you'll need to either:
1. implement this as the standard describes, with unit tests (not `test-harness` integration tests) to validate that it works with a variety of input, OR
2. rely on a library which parses multipart data according to the standard.

ilanashapiro replied:
Hi Ed, thank you so much for your feedback. I discovered a library called parse-multiform-data which I'm installing as an npm package that seemed to be the most reliable method. I made changes according to this, such as parsing the boundary parameter from the Content-Type header field in order to pass it to the library for parsing the body. I left some comments on the recent changes that hopefully help clarify some of the decisions I made when implementing this. Thanks again and please let me know if you'd like any other changes!

EdVinyard reviewed (changes requested):
We really appreciate your update, but I think there are some cases the `parseMultipartFormDataParams()` function doesn't yet handle regarding the boundary between multipart entities.

ilanashapiro commented on packages/http/src/validator/index.ts:
```diff
@@ -75,17 +76,22 @@ const validateInputBody = (
+      // parse boundary string from content-type in case media type is multipart/form-data
+      const multipart = require('parse-multipart-data');
+      const contentTypeHeader = headers.get('content-type');
+      const multipartBoundary = contentTypeHeader ? multipart.getBoundary(contentTypeHeader) : "";
```
Here, we parse the boundary parameter value from the Content-Type header field as specified by https://datatracker.ietf.org/doc/html/rfc2046#section-5.1.1. This is later passed into the parse-multipart-data npm package to correctly parse the body.

ilanashapiro commented on packages/http/src/mocker/index.ts:
```diff
@@ -137,19 +137,25 @@ function runCallbacks({
+  
+  // parse boundary string from content-type in case media type is multipart/form-data
+  const multipart = require('parse-multipart-data');
+  const multipartBoundary = multipart.getBoundary(contentTypeHeader);
```
Also here, we parse the boundary parameter value from the Content-Type header field as specified by https://datatracker.ietf.org/doc/html/rfc2046#section-5.1.1. This is later passed into the parse-multipart-data npm package to correctly parse the body in the function parseMultipartFormDataParams.

ilanashapiro commented on packages/http/src/validator/index.ts:
```diff
@@ -75,17 +76,22 @@ const validateInputBody = (
+      const multipart = require('parse-multipart-data');
+      const contentTypeHeader = headers.get('content-type');
+      const multipartBoundary = contentTypeHeader ? multipart.getBoundary(contentTypeHeader) : "";
+      const mediaType = contentTypeHeader ? contentTypeHeader.replace(new RegExp(";\\s*boundary=" + multipartBoundary), "") : contentTypeHeader;
```
After end-end and unit testing, the content-type header appeared in both formats: "multipart/form-data;boundary=--...." and "multipart/form-data; boundary=--...." (with the space after the semicolon). Thus, I'm handling possible whitespace between the media type and the boundary string.

EdVinyard replied:
1. I'm concerned about your use of a regular expression here.  Notice that the boundary may contain any 7-bit ASCII character, which I think includes many characters with special meanings inside a regular expression (e.g., `.`, `*`, `(`).  Consider using a library to parse the Content-Type header in its entirety.  Here's a very simple example using [whatwg-mimetype](https://github.com/jsdom/whatwg-mimetype):

	```TypeScript
	import MIMEType from 'whatwg-mimetype';
	
	const t = new MIMEType('multipart/form-data; boundary=.*.?.+zzz');
	console.log(t?.essence);
	console.log(t.parameters.get('boundary'));
	```

3. It looks like this block of code is a duplicate of what you added in `mocker/index.ts`.  Could it be factored out?

ilanashapiro commented on packages/http/src/mocker/index.ts:
```diff
@@ -137,19 +137,25 @@ function runCallbacks({
+  // parse boundary string from content-type in case media type is multipart/form-data
+  const multipart = require('parse-multipart-data');
+  const multipartBoundary = multipart.getBoundary(contentTypeHeader);
+  const mediaType = contentTypeHeader.replace(new RegExp(";\\s*boundary=" + multipartBoundary), "");
```
See comment in packages/http/src/validator/index.ts

ilanashapiro commented on packages/http/src/validator/validators/body.ts:
```diff
@@ -54,6 +55,30 @@ export function splitUriParams(target: string) {
+  const multipart = require('parse-multipart-data');
+  const bufferBody = Buffer.from(target, "utf-8");
+
+  // the parse-multipart-data package requires that the body is passed in as a buffer, not a string
```
See: https://github.com/nachomazzara/parse-multipart-data/issues/14#:~:text=I%20was%20having%20a%20similar%20problem%20with%20a%20different%20payload.%20I%20opened%20up%20the%20dist/%20and%20realized%20that%20the%20author%20named%20the%20body%20param%20multipartBodyBuffer.%20I%20was%20passing%20the%20body%20as%20a%20string%2C%20but%20when%20I%20changed%20to%20pass%20it%20as%20a%20Buffer%20I%20got%20results!

And also the source code: https://github.com/nachomazzara/parse-multipart-data/blob/master/src/multipart.ts#L34C80-L34C80

ilanashapiro commented on packages/http/src/validator/validators/body.ts:
```diff
@@ -54,6 +55,30 @@ export function splitUriParams(target: string) {
+  const parts = multipart.parse(bufferBody, multipartBoundary);
+
+  return parts.reduce((result: Dictionary<string>, pair: string) => {
+    result[pair['name']] = pair['data'].toString();
```
then convert back from Buffer to String for further processing by Prism

ilanashapiro commented on packages/http/src/validator/validators/body.ts:
```diff
@@ -54,6 +55,30 @@ export function splitUriParams(target: string) {
 
+export function parseMultipartFormDataParams(target: string, multipartBoundary?: string) {
+  if(!multipartBoundary) {
+    const error = "Boundary parameter for multipart/form-data is not defined or generated in the request header. Try removing manually defined content-type from your request header if it exists.";
```
The boundary string MUST be non-empty if this is multipart/form-data (which it is if we're in this function). Thus, we return with error. According to https://bobbyhadz.com/blog/error-multipart-boundary-not-found-in-express-js, https://github.com/postmanlabs/postman-app-support/issues/191, https://stackoverflow.com/questions/40351429/formdata-how-to-get-or-set-boundary-in-multipart-form-data-angular, this can occur when the user is manually setting the content-type, which means the server is not generating it (and the boundary string). Thus, the possible fix is suggested in the error message.

ilanashapiro replied:
(Returning an error seemed to me a better solution than just returning an empty dictionary as it keeps the user more informed. Thus, we the return type is the Either monad).

EdVinyard commented on packages/http/src/validator/index.ts:
```diff
@@ -75,17 +76,22 @@ const validateInputBody = (
     E.map(b => [...b, caseless(headers || {})] as const),
     E.chain(([requestBody, body, headers]) => {
+      // parse boundary string from content-type in case media type is multipart/form-data
+      const multipart = require('parse-multipart-data');
```
We'd prefer to use `import` statements rather than dynamic `require` whenever possible.

ilanashapiro commented on packages/http/src/validator/validators/body.ts:
```diff
@@ -47,11 +49,35 @@ export function deserializeFormBody(
 
 export function splitUriParams(target: string) {
-  return target.split('&').reduce((result: Dictionary<string>, pair: string) => {
+  return E.right(target.split('&').reduce((result: Dictionary<string>, pair: string) => {
```
Lifting to Either monad in order to match the return type of parseMultipartFormDataParams, which is of type Either as it may return an error.

ilanashapiro commented on packages/http/src/validator/validators/body.ts:
```diff
@@ -70,12 +96,12 @@ export function findContentByMediaTypeOrFirst(specs: IMediaTypeContent[], mediaT
 
   return pipe(
-    validateAgainstReservedCharacters(encodedUriParams, encodings, prefix),
+    content.mediaType === "multipart/form-data" ? parseMultipartFormDataParams(target, multipartBoundary) : splitUriParams(target),
```
Since parseMultipartFormDataParams returns E.left if the boundary string is empty, the return type is Either, and I wrapped the result of splitUriParams in the Either monad as well to match the return type of parseMultipartFormDataParams. Thus, we move this statement into the expression pipeline.

ilanashapiro commented on packages/http/src/mocker/index.ts:
```diff
@@ -137,19 +139,24 @@ function runCallbacks({
+  
+  const requestBody = request.body as string;
+  const encodedUriParams = pipe(
+    mediaType === "multipart/form-data" ? parseMultipartFormDataParams(requestBody, multipartBoundary) : splitUriParams(requestBody),
```
parseMultipartFormDataParams returns an error wrapped in E.Left if the boundary sting is empty. Here I convert that to a default value (i.e. empty dictionary) if that occurs. I'm not sure if I'm supposed to be doing anything else here -- this same behavior (unwrapping the monad using defaults) is implemented for findContentByMediaTypeOrFirst just below on line 169, so I attempted to emulate that.

ilanashapiro commented:
Hi Ed, thanks for your feedback, I am now doing all Content-Type header parsing using the whatwg-mimetype library you recommended, and the dynamic require statements were changed to imports. As for the duplicate code block, I moved it to a small utility function called parseMIMEHeader I created in validators/headers.ts. Finally, I also attempted to improve the error reporting for empty multipart boundary strings by wrapping parseMultipartFormDataParams in the Either monad and returning a descriptive error (which meant I also needed to wrap splitUriParams in the Either monad to maintain equivalent types).

chohmann reviewed (approved):
Thank you for your contribution and for working through our reviews with us!

chohmann commented:
This is included in version 5.2.0 of Prism.

rattrayalex commented:
Hooray, thanks Ilana, Ed, and Chelsea!
